File: server/routes/folder-import.js
const express = require("express");
const multer = require("multer");
const { authenticateToken } = require("../middleware/auth");
const FileStorageService = require("../services/FileStorageService");

const router = express.Router();

// Configure multer for memory storage (process files in memory)
const upload = multer({
  storage: multer.memoryStorage(),
  limits: {
    fileSize: 50 * 1024 * 1024, // 50MB max file size
    files: 1000, // Max 1000 files per request (increased from 100)
    fieldSize: 10 * 1024 * 1024 // 10MB field size limit
  },
  fileFilter: (req, file, cb) => {
    // Accept all file types for now
    cb(null, true);
  }
});

// POST /api/folder-import - Import folder with JSON and binary files
router.post("/", authenticateToken, upload.array('files'), async (req, res) => {
  try {
    console.log("Folder import request received");
    const files = req.files;
    const { folderName } = req.body;

    console.log(`Processing ${files ? files.length : 0} files for folder: ${folderName}`);

    if (!files || files.length === 0) {
      console.log("No files uploaded");
      return res.status(400).json({ error: 'No files uploaded' });
    }

    // Log file count and sizes for debugging
    const totalSize = files.reduce((sum, file) => sum + file.size, 0);
    console.log(`Total files: ${files.length}, Total size: ${Math.round(totalSize / 1024)}KB`);

    const fileStorage = new FileStorageService();
    const result = await processFolderImport(files, folderName, fileStorage);

    console.log(`Import completed: ${result.jsonFiles} JSON files, ${result.binaryFiles} binary files, ${result.errors.length} errors`);

    res.json({
      success: true,
      message: `Successfully processed ${result.totalFiles} files (${result.jsonFiles} JSON, ${result.binaryFiles} binary)`,
      data: result.structure,
      stats: {
        totalFiles: result.totalFiles,
        jsonFiles: result.jsonFiles,
        binaryFiles: result.binaryFiles,
        errors: result.errors.length,
        errorDetails: result.errors
      }
    });

  } catch (error) {
    console.error("Folder import error:", error);

    // Handle specific multer errors
    if (error.code === 'LIMIT_FILE_COUNT') {
      return res.status(400).json({
        success: false,
        error: "Too many files in upload",
        details: `Maximum ${upload.options?.limits?.files || 1000} files allowed`
      });
    }

    if (error.code === 'LIMIT_FILE_SIZE') {
      return res.status(400).json({
        success: false,
        error: "File too large",
        details: "Maximum file size is 50MB"
      });
    }

    res.status(500).json({
      success: false,
      error: "Failed to import folder",
      details: error.message
    });
  }
});

// Handle multer errors specifically
router.use((error, req, res, next) => {
  if (error instanceof multer.MulterError) {
    console.error("Multer error:", error);

    if (error.code === 'LIMIT_FILE_COUNT') {
      return res.status(400).json({
        success: false,
        error: "Too many files",
        details: "Maximum 1000 files allowed per upload"
      });
    }

    if (error.code === 'LIMIT_FILE_SIZE') {
      return res.status(400).json({
        success: false,
        error: "File too large",
        details: "Maximum file size is 50MB"
      });
    }

    return res.status(400).json({
      success: false,
      error: "File upload error",
      details: error.message
    });
  }

  next(error);
});

/**
 * Process folder import with both JSON and binary files
 * @param {Array} files - Array of uploaded files
 * @param {string} folderName - Name of the root folder
 * @param {FileStorageService} fileStorage - File storage service instance
 * @returns {Object} Processing result with structure and stats
 */
async function processFolderImport(files, folderName, fileStorage) {
  // Two-stage import:
  // 1) Import JSON files only and build the folder structure with parsed JSON preserved
  // 2) Import non-JSON (binary) files and attach file references into the already-built structure
  const structure = {};
  const errors = [];
  let jsonFiles = 0;
  let binaryFiles = 0;

  // Helper to ensure the nested folder path exists in the structure and return the leaf object
  function ensureFolderPath(root, parts) {
    let current = root;
    for (const part of parts) {
      if (!current[part] || typeof current[part] !== 'object' || Array.isArray(current[part])) {
        current[part] = {};
      }
      current = current[part];
    }
    return current;
  }

  // First pass: process JSON files only
  for (const file of files) {
    try {
      const relativePath = file.originalname;
      const pathParts = relativePath.split('/').filter(part => part.length > 0);

      if (pathParts.length === 0) {
        errors.push({ file: file.originalname, error: 'Invalid file path' });
        continue;
      }

      const fileName = pathParts[pathParts.length - 1];
      const fileExtension = getFileExtension(fileName);
      if (fileExtension !== '.json') continue;

      const folderParts = pathParts.slice(0, -1);
      const parent = ensureFolderPath(structure, folderParts);

      try {
        const fileContent = file.buffer.toString('utf8');
        const jsonContent = JSON.parse(fileContent);
        const propertyName = fileName.replace(/\.json$/i, '');
        parent[propertyName] = jsonContent;
        jsonFiles++;
      } catch (parseError) {
        errors.push({ file: file.originalname, error: `JSON parse error: ${parseError.message}` });
      }
    } catch (error) {
      errors.push({ file: file.originalname, error: error.message });
    }
  }

  // Second pass: process non-JSON (binary) files and attach them into the existing structure
  for (const file of files) {
    try {
      const relativePath = file.originalname;
      const pathParts = relativePath.split('/').filter(part => part.length > 0);

      if (pathParts.length === 0) {
        // Already logged in first pass if invalid
        continue;
      }

      const fileName = pathParts[pathParts.length - 1];
      const fileExtension = getFileExtension(fileName);
      if (fileExtension === '.json') continue;

      const folderParts = pathParts.slice(0, -1);
      const parent = ensureFolderPath(structure, folderParts);

      // Determine a safe property name for the file inside the parent folder
      let propertyName = getFileNameWithoutExtension(fileName);
      if (parent.hasOwnProperty(propertyName)) {
        // Avoid clobbering existing properties (e.g. a JSON with same name)
        let idx = 1;
        while (parent.hasOwnProperty(`${propertyName}_${idx}`)) idx++;
        propertyName = `${propertyName}_${idx}`;
      }

      try {
        const fileMetadata = await fileStorage.storeFile(
          file.buffer,
          fileName,
          file.mimetype || 'application/octet-stream'
        );

        const downloadUrl = await fileStorage.generateDownloadUrl(fileMetadata);

        parent[propertyName] = {
          _type: 'file',
          _metadata: {
            originalName: fileName,
            mimeType: file.mimetype || 'application/octet-stream',
            size: file.size,
            storageKey: fileMetadata.fileId,
            storageType: fileMetadata.storageType
          },
          _link: downloadUrl
        };

        binaryFiles++;
      } catch (storageError) {
        errors.push({ file: file.originalname, error: `Storage error: ${storageError.message}` });
      }
    } catch (error) {
      errors.push({ file: file.originalname, error: error.message });
    }
  }

  return {
    structure,
    totalFiles: files.length,
    jsonFiles,
    binaryFiles,
    errors
  };
}

/**
 * Get file extension
 */
function getFileExtension(filename) {
  const lastDotIndex = filename.lastIndexOf('.');
  return lastDotIndex === -1 ? '' : filename.substring(lastDotIndex).toLowerCase();
}

/**
 * Get filename without extension
 */
function getFileNameWithoutExtension(filename) {
  const lastDotIndex = filename.lastIndexOf('.');
  return lastDotIndex === -1 ? filename : filename.substring(0, lastDotIndex);
}

module.exports = router;

File: server-go/internal/http/handlers/folder_import.go
package handlers

import (
	"encoding/json"
	"fmt"
	"mime/multipart"
	"net/http"
	"path/filepath"
	"strings"

	"github.com/gin-gonic/gin"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"

	"your.module/config-manager/internal/files"
	"your.module/config-manager/internal/logger"
	"your.module/config-manager/internal/mongo"
	"your.module/config-manager/internal/types"
)

type FolderImportHandler struct {
	storage *files.StorageManager
	db      *mongo.Client
	log     *logger.Logger
}

func NewFolderImportHandler(storage *files.StorageManager, db *mongo.Client, log *logger.Logger) *FolderImportHandler {
	return &FolderImportHandler{
		storage: storage,
		db:      db,
		log:     log,
	}
}

func (h *FolderImportHandler) Register(api *gin.RouterGroup) {
	api.POST("/folder-import", h.importFolder)
}

// POST /api/folder-import - Import folder with JSON and binary files
func (h *FolderImportHandler) importFolder(c *gin.Context) {
	// Parse multipart form (max 1000 files)
	err := c.Request.ParseMultipartForm(32 << 20) // 32 MB max memory
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"success": false,
			"error":   "Failed to parse multipart form",
			"details": err.Error(),
		})
		return
	}

	form := c.Request.MultipartForm
	if form == nil || form.File["files"] == nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"success": false,
			"error":   "No files uploaded",
		})
		return
	}

	uploadedFiles := form.File["files"]
	folderName := c.PostForm("folderName")
	configID := c.PostForm("configId")
	propertyPath := c.PostForm("propertyPath")

	// Get relative paths (can be multiple form values)
	relativePaths := c.Request.Form["relativePaths"]

	fmt.Printf("Folder import: %d files, folder=%s, configId=%s, propertyPath=%s\n",
		len(uploadedFiles), folderName, configID, propertyPath)

	// Process folder import
	result, err := h.processFolderImport(c, uploadedFiles, folderName, relativePaths)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"success": false,
			"error":   "Failed to process folder import",
			"details": err.Error(),
		})
		return
	}

	// If configId and propertyPath provided, attach to configuration
	if configID != "" && propertyPath != "" {
		err := h.attachToConfiguration(c, configID, propertyPath, result.Structure)
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{
				"success": false,
				"error":   "Import succeeded but failed to attach to configuration",
				"details": err.Error(),
				"data":    result.Structure,
				"stats": gin.H{
					"totalFiles":   result.TotalFiles,
					"jsonFiles":    result.JSONFiles,
					"binaryFiles":  result.BinaryFiles,
					"errors":       len(result.Errors),
					"errorDetails": result.Errors,
				},
			})
			return
		}

		c.JSON(http.StatusOK, gin.H{
			"success":  true,
			"message":  fmt.Sprintf("Successfully imported and attached %d files to property \"%s\"", result.TotalFiles, propertyPath),
			"attached": true,
			"configId": configID,
			"propertyPath": propertyPath,
			"stats": gin.H{
				"totalFiles":   result.TotalFiles,
				"jsonFiles":    result.JSONFiles,
				"binaryFiles":  result.BinaryFiles,
				"errors":       len(result.Errors),
				"errorDetails": result.Errors,
			},
		})
		return
	}

	// Normal import without attachment
	c.JSON(http.StatusOK, gin.H{
		"success":  true,
		"message":  fmt.Sprintf("Successfully processed %d files (%d JSON, %d binary)", result.TotalFiles, result.JSONFiles, result.BinaryFiles),
		"data":     result.Structure,
		"attached": false,
		"stats": gin.H{
			"totalFiles":   result.TotalFiles,
			"jsonFiles":    result.JSONFiles,
			"binaryFiles":  result.BinaryFiles,
			"errors":       len(result.Errors),
			"errorDetails": result.Errors,
		},
	})
}

type ImportResult struct {
	Structure   map[string]interface{}
	TotalFiles  int
	JSONFiles   int
	BinaryFiles int
	Errors      []map[string]string
}

func (h *FolderImportHandler) processFolderImport(c *gin.Context, uploadedFiles []*multipart.FileHeader, folderName string, relativePaths []string) (*ImportResult, error) {
	structure := make(map[string]interface{})
	var errors []map[string]string
	jsonFiles := 0
	binaryFiles := 0

	// First pass: process JSON files only
	for idx, fileHeader := range uploadedFiles {
		file, err := fileHeader.Open()
		if err != nil {
			errors = append(errors, map[string]string{
				"file":  fileHeader.Filename,
				"error": fmt.Sprintf("Failed to open file: %v", err),
			})
			continue
		}

		// Read file content
		fileContent := make([]byte, fileHeader.Size)
		_, err = file.Read(fileContent)
		file.Close()
		if err != nil {
			errors = append(errors, map[string]string{
				"file":  fileHeader.Filename,
				"error": fmt.Sprintf("Failed to read file: %v", err),
			})
			continue
		}

		// Get relative path
		var relativePath string
		if idx < len(relativePaths) && relativePaths[idx] != "" {
			relativePath = relativePaths[idx]
		} else {
			relativePath = fileHeader.Filename
		}

		// Normalize path
		relativePath = strings.ReplaceAll(relativePath, "\\", "/")
		relativePath = strings.TrimPrefix(relativePath, "/")
		pathParts := strings.Split(relativePath, "/")
		pathParts = filterEmpty(pathParts)

		// Remove folder name if it's the first part
		if folderName != "" && len(pathParts) > 0 && pathParts[0] == folderName {
			pathParts = pathParts[1:]
		}

		if len(pathParts) == 0 {
			errors = append(errors, map[string]string{
				"file":  fileHeader.Filename,
				"error": "Invalid file path",
			})
			continue
		}

		fileName := pathParts[len(pathParts)-1]
		fileExt := strings.ToLower(filepath.Ext(fileName))

		// Only process JSON files in first pass
		if fileExt != ".json" {
			continue
		}

		folderParts := pathParts[:len(pathParts)-1]
		parent := ensureFolderPath(structure, folderParts)

		// Parse JSON
		var jsonContent interface{}
		err = json.Unmarshal(fileContent, &jsonContent)
		if err != nil {
			errors = append(errors, map[string]string{
				"file":  fileHeader.Filename,
				"error": fmt.Sprintf("JSON parse error: %v", err),
			})
			continue
		}

		propertyName := getSanitizedFileNameWithoutExtension(fileName)
		if parentMap, ok := parent.(map[string]interface{}); ok {
			parentMap[propertyName] = jsonContent
			jsonFiles++
		}
	}

	// Second pass: process binary files
	for idx, fileHeader := range uploadedFiles {
		file, err := fileHeader.Open()
		if err != nil {
			continue // Already logged in first pass
		}

		// Read file content
		fileContent := make([]byte, fileHeader.Size)
		_, err = file.Read(fileContent)
		file.Close()
		if err != nil {
			continue // Already logged
		}

		// Get relative path
		var relativePath string
		if idx < len(relativePaths) && relativePaths[idx] != "" {
			relativePath = relativePaths[idx]
		} else {
			relativePath = fileHeader.Filename
		}

		// Normalize path
		relativePath = strings.ReplaceAll(relativePath, "\\", "/")
		relativePath = strings.TrimPrefix(relativePath, "/")
		pathParts := strings.Split(relativePath, "/")
		pathParts = filterEmpty(pathParts)

		// Remove folder name if it's the first part
		if folderName != "" && len(pathParts) > 0 && pathParts[0] == folderName {
			pathParts = pathParts[1:]
		}

		if len(pathParts) == 0 {
			continue // Already logged
		}

		fileName := pathParts[len(pathParts)-1]
		fileExt := strings.ToLower(filepath.Ext(fileName))

		// Skip JSON files (already processed)
		if fileExt == ".json" {
			continue
		}

		folderParts := pathParts[:len(pathParts)-1]
		parent := ensureFolderPath(structure, folderParts)

		// Store binary file
		mimeType := fileHeader.Header.Get("Content-Type")
		if mimeType == "" {
			mimeType = "application/octet-stream"
		}

		fileMetadata, err := h.storage.StoreFile(c.Request.Context(), fileContent, fileName, mimeType)
		if err != nil {
			errors = append(errors, map[string]string{
				"file":  fileHeader.Filename,
				"error": fmt.Sprintf("Storage error: %v", err),
			})
			continue
		}

		downloadURL, err := h.storage.GenerateDownloadURL(c.Request.Context(), fileMetadata)
		if err != nil {
			downloadURL = ""
		}

		// Create file object
		fileObject := map[string]interface{}{
			"_type": "file",
			"_metadata": map[string]interface{}{
				"originalName": fileName,
				"mimeType":     mimeType,
				"size":         fileHeader.Size,
				"storageKey":   fileMetadata.FileID,
				"storageType":  fileMetadata.StorageType,
			},
			"_link": downloadURL,
		}

		propertyName := getSanitizedFileNameWithoutExtension(fileName)
		if parentMap, ok := parent.(map[string]interface{}); ok {
			// Avoid collision with existing properties
			originalName := propertyName
			counter := 1
			for {
				if _, exists := parentMap[propertyName]; !exists {
					break
				}
				propertyName = fmt.Sprintf("%s_%d", originalName, counter)
				counter++
			}
			parentMap[propertyName] = fileObject
			binaryFiles++
		}
	}

	return &ImportResult{
		Structure:   structure,
		TotalFiles:  len(uploadedFiles),
		JSONFiles:   jsonFiles,
		BinaryFiles: binaryFiles,
		Errors:      errors,
	}, nil
}

func (h *FolderImportHandler) attachToConfiguration(c *gin.Context, configID, propertyPath string, structure map[string]interface{}) error {
	oid, err := primitive.ObjectIDFromHex(configID)
	if err != nil {
		return fmt.Errorf("invalid configId: %v", err)
	}

	// Verify configuration exists
	var config types.Configuration
	err = h.db.Configurations.FindOne(c, bson.M{"_id": oid}).Decode(&config)
	if err != nil {
		return fmt.Errorf("configuration not found: %v", err)
	}

	// Build update with structure at property path
	dataUpdate := make(map[string]interface{})
	setValueAtPath(dataUpdate, propertyPath, structure)

	// Update configuration
	update := bson.M{
		"$set": bson.M{
			"data": dataUpdate,
		},
	}

	_, err = h.db.Configurations.UpdateOne(c, bson.M{"_id": oid}, update)
	if err != nil {
		return fmt.Errorf("failed to update configuration: %v", err)
	}

	return nil
}

// Helper functions

func ensureFolderPath(root interface{}, parts []string) interface{} {
	current := root
	for _, part := range parts {
		if currentMap, ok := current.(map[string]interface{}); ok {
			if _, exists := currentMap[part]; !exists {
				currentMap[part] = make(map[string]interface{})
			}
			current = currentMap[part]
		}
	}
	return current
}

func getSanitizedFileNameWithoutExtension(filename string) string {
	// Remove extension
	name := strings.TrimSuffix(filename, filepath.Ext(filename))
	// Replace dots with underscores
	name = strings.ReplaceAll(name, ".", "_")
	return name
}

func filterEmpty(parts []string) []string {
	var result []string
	for _, part := range parts {
		if part != "" {
			result = append(result, part)
		}
	}
	return result
}

